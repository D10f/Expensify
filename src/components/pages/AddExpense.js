import React from 'react'
import { connect } from 'react-redux'
// import { addExpense } from '../../redux/actions/expenses'
import { database } from '../../firebase/firebase'
import ExpenseForm from '../ExpenseForm'

export const AddExpensePage = (props) => {

  const onSubmit = (expenseData) => {
    database.ref('expenses').push(expenseData).then(() => props.history.push('/'))
    // props.addExpense(expense)
  }

  return (
    <section>
      <h3>Create New Expense</h3>
      <ExpenseForm onSubmit={onSubmit} />
    </section>
  )
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
