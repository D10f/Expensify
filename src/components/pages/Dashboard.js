import React from 'react'
import ExpenseSummary from '../ExpenseSummary'
import ExpenseList from '../ExpenseList'
import ExpenseListFilters from '../ExpenseListFilters'

export const Dashboard = () => (
  <section>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </section>
)

export default Dashboard
