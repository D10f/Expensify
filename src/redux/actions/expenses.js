import { v4 as uuid } from 'uuid'

export const addExpense = ({ id = uuid(), description = '', notes = '', amount = 100, createdAt = 0 } = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id,
    description,
    notes,
    amount,
    createdAt
  }
})

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const editExpense = (id, updates = {}) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})
