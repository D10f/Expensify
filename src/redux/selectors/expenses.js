import moment from 'moment'

export default (expenses, { text, startDate, endDate, sortBy }) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true
    const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day') : true
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt > startDate
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt < endDate
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    }

    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
}
