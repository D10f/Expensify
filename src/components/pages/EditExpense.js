import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../../redux/actions/expenses'
import { database } from '../../firebase/firebase'
import ExpenseForm from '../ExpenseForm'

export class EditExpensePage extends React.Component {

  onClick = () => {
    database.ref(`expenses/${this.props.expense.id}`).remove().then(() => this.props.history.push('/'))
    // this.props.removeExpense({ id: this.props.expense.id })
  }

  onSubmit = (updatedExpense) => {
    database.ref(`expenses/${this.props.expense.id}`).update(updatedExpense).then(() => this.props.history.push('/'))
    // this.props.editExpense(this.props.expense.id, updatedExpense)
  }

  render(){
    return (
      <section>
        <button onClick={this.onClick}>Delete Expense</button>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
      </section>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
}
//
// const mapDispatchToProps = (dispatch) => ({
//   editExpense: (id, updates) => dispatch(editExpense(id, updates)),
//   removeExpense: (id) => dispatch(removeExpense(id))
// })

export default connect(mapStateToProps)(EditExpensePage)
