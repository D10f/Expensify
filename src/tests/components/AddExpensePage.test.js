import React from 'react'
import { shallow } from 'enzyme'
import { AddExpensePage } from '../../components/pages/AddExpense'
import expenses from '../fixtures/expenses'

let startAddExpense, history, wrapper

beforeEach(() => {
  startAddExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={history} />)
})

test('Should render correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should fire onSubmit correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
  expect(startAddExpense).toHaveBeenCalledWith(expenses[1])
  expect(history.push).toHaveBeenCalledWith('/')
})
