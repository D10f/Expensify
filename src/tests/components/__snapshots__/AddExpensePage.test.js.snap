// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<section>
  <h3>
    Create New Expense
  </h3>
  <ExpenseForm
    onSubmit={[Function]}
  />
</section>
`;
